controllers/integration/integration_adapter.go

```diff
@@ -137,8 +222,167 @@ func (a *Adapter) getAllApplicationSnapshots() (*[]appstudioshared.ApplicationSn
+				integrationPipelineRuns = append(integrationPipelineRuns, *integrationPipelineRun)
+			}
+		} else {
+			integrationPipelineRuns = append(integrationPipelineRuns, *a.pipelineRun)
+			a.logger.Info("The current integrationPipelineRun matches the integration test scenario",
+				"IntegrationTestScenario.Name", integrationTestScenario.Name,
+				"integrationPipelineRun.Name", a.pipelineRun.Name)
+		}
+	}
+
+	return &integrationPipelineRuns, nil
+}
 
+// getLatestPipelineRunForApplicationSnapshotAndScenario returns the latest Integration PipelineRun for the
+// associated ApplicationSnapshot and IntegrationTestScenario. In the case the List operation fails,
+// an error will be returned.
+func (a *Adapter) getLatestPipelineRunForApplicationSnapshotAndScenario(applicationSnapshot *appstudioshared.ApplicationSnapshot, integrationTestScenario *v1alpha1.IntegrationTestScenario) (*tektonv1beta1.PipelineRun, error) {
+	integrationPipelineRuns := &tektonv1beta1.PipelineRunList{}
+	var latestIntegrationPipelineRun = &tektonv1beta1.PipelineRun{}
+	opts := []client.ListOption{
+		client.InNamespace(a.application.Namespace),
+		client.MatchingLabels{
+			"pipelines.appstudio.openshift.io/type": "test",
+			"test.appstudio.openshift.io/snapshot":  applicationSnapshot.Name,
+			"test.appstudio.openshift.io/scenario":  integrationTestScenario.Name,
+		},
+	}
+
+	err := a.client.List(a.context, integrationPipelineRuns, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	latestIntegrationPipelineRun = nil
+	for _, pipelineRun := range integrationPipelineRuns.Items {
+		if pipelineRun.Status.GetCondition(apis.ConditionSucceeded).IsTrue() {
+			if latestIntegrationPipelineRun == nil {
+				latestIntegrationPipelineRun = &pipelineRun
+			} else {
+				if pipelineRun.Status.CompletionTime.Time.After(latestIntegrationPipelineRun.Status.CompletionTime.Time) {
+					latestIntegrationPipelineRun = &pipelineRun
+				}
+			}
+		}
+	}
+	if latestIntegrationPipelineRun != nil {
+		return latestIntegrationPipelineRun, nil
+	}
+
+	return nil, err
+}
+
+// calculateIntegrationPipelineRunOutcome checks the Tekton results for a PipelineRun and calculates the overall outcome.
+// If any of the tasks with the HACBS_TEST_OUTPUT result don't have the `result` field set to SUCCESS/SKIPPED,
+// it returns false
+func (a *Adapter) calculateIntegrationPipelineRunOutcome(pipelineRun *tektonv1beta1.PipelineRun) (bool, error) {
+	for _, taskRun := range pipelineRun.Status.TaskRuns {
+		for _, taskRunResult := range taskRun.Status.TaskRunResults {
+			if taskRunResult.Name == "HACBS_TEST_OUTPUT" {
+				var testOutput map[string]interface{}
+				err := json.Unmarshal([]byte(taskRunResult.Value), &testOutput)
+				if err != nil {
+					return false, err
+				}
+				a.logger.Info("Found a task with HACBS_TEST_OUTPUT",
+					"taskRun.Name", taskRun.PipelineTaskName,
+					"taskRun Result", testOutput["result"])
+				if testOutput["result"] != "SUCCESS" && testOutput["result"] != "SKIPPED" {
```

Comment:
Should `&&` be changed to `||`?

Reply:
Sorry, I am wrong, you are right.

Reply:
No need to apologize, the condition is kinda hard to read, I'll see if I can make it a bit more readable.

Reply:
Changed the condition to be more readable and added an explanatory comment above it

Reply:
Thank you. It is clearer now.